runtime/js/marshal.js

```diff
@@ -630,98 +654,113 @@ function caml_marshal_data_size(s, ofs) {
 }
 
 //Provides: MlObjectTable
-var MlObjectTable;
-if (typeof globalThis.Map === "undefined") {
-  MlObjectTable = (function () {
-    /* polyfill (using linear search) */
-    function NaiveLookup(objs) {
-      this.objs = objs;
+class BaseMlObjectTable {
+  constructor() {
+    this.objs = [];
+  }
+
+  store(v) {
+    this.lookup.set(v, this.objs.length);
+    this.objs.push(v);
+  }
+
+  recall(v) {
+    var i = this.lookup.get(v);
+    return i === undefined
+      ? undefined
+      : this.objs.length - i; /* index is relative */
+  }
+}
+// polyfill (using linear search)
+class NaiveLookup {
+  constructor(objs) {
+    this.objs = objs;
+  }
+
+  get(v) {
+    for (let i = 0; i < this.objs.length; i++) {
+      if (this.objs[i] === v) return i;
     }
-    NaiveLookup.prototype.get = function (v) {
-      for (var i = 0; i < this.objs.length; i++) {
-        if (this.objs[i] === v) return i;
-      }
-    };
-    NaiveLookup.prototype.set = function () {
-      // Do nothing here. [MlObjectTable.store] will push to [this.objs] directly.
-    };
+  }
 
-    return function MlObjectTable() {
-      this.objs = [];
+  set() {
+    // Do nothing here. [MlObjectTable.store] will push to [this.objs] directly.
+  }
+}
+var MlObjectTable;
+if (typeof globalThis.Map === "undefined") {
```

Comment:
If we start using `class`,  we might  as well expect `Map` to be available. Would you mind opening a separate PR for this ? and rebase this PR on top ?

Reply:
Sure

Reply:
I will rebase this PR later.

Reply:
Rebased!